src/lib/planviz/abstract_visualizer.hpp

```diff
@@ -0,0 +1,156 @@
+#include <boost/graph/graphviz.hpp>
+#include <string>
+#include <unordered_map>
+#include <utility>
+
+#include "operators/print.hpp"
+#include "planviz/graphviz_config.hpp"
+
+namespace opossum {
+
+struct GraphvizConfig {
+  std::string layout = GraphvizLayout::Dot;
+  std::string format = GraphvizRenderFormat::Png;
+};
+
+struct VizGraphInfo {
+  std::string bg_color = GraphvizColor::Transparent;
+  std::string rankdir = "BT";
+  float ratio = 0.5f;
+};
+
+struct VizVertexInfo {
+  std::string label;
+  std::string color = GraphvizColor::White;
+  std::string font_color = GraphvizColor::White;
+  std::string shape = GraphvizShape::Rectangle;
+  uint8_t pen_width = 1u;
+};
+
+struct VizEdgeInfo {
+  std::string label;
+  std::string color = GraphvizColor::White;
+  std::string font_color = GraphvizColor::White;
+  uint8_t pen_width = 1u;
+};
+
+template <typename GraphBase>
+class AbstractVisualizer {
+  //                                  Edge list    Vertex list   Directed graph
+  using Graph = boost::adjacency_list<boost::vecS, boost::vecS, boost::directedS,
+                                      // Vertex info Edge info    Graph info
+                                      VizVertexInfo, VizEdgeInfo, VizGraphInfo>;
+
+ public:
+  AbstractVisualizer() : AbstractVisualizer(GraphvizConfig{}, VizGraphInfo{}, VizVertexInfo{}, VizEdgeInfo{}) {}
+
+  AbstractVisualizer(GraphvizConfig graphviz_config, VizGraphInfo graph_info, VizVertexInfo vertex_info,
+                     VizEdgeInfo edge_info)
+      : _graphviz_config(std::move(graphviz_config)),
+        _graph_info(std::move(graph_info)),
+        _default_vertex(std::move(vertex_info)),
+        _default_edge(std::move(edge_info)) {
+    // Add global Graph properties
+    _add_graph_property("rankdir", _graph_info.rankdir);
+    _add_graph_property("bgcolor", _graph_info.bg_color);
+    _add_graph_property("ratio", _graph_info.ratio);
+
+    // Add vertex properties
+    _add_property("node_id", boost::vertex_index);
+    _add_property("color", &VizVertexInfo::color);
+    _add_property("label", &VizVertexInfo::label);
+    _add_property("shape", &VizVertexInfo::shape);
+    _add_property("fontcolor", &VizVertexInfo::font_color);
+    _add_property("penwidth", &VizVertexInfo::pen_width);
+
+    // Add edge properties
+    _add_property("color", &VizEdgeInfo::color);
+    _add_property("fontcolor", &VizEdgeInfo::font_color);
+    _add_property("label", &VizEdgeInfo::label);
+    _add_property("penwidth", &VizEdgeInfo::pen_width);
+  }
+
+  void visualize(const GraphBase& graph_base, const std::string& graph_filename, const std::string& img_filename) {
+    _build_graph(graph_base);
+    std::ofstream file(graph_filename);
+    boost::write_graphviz_dp(file, _graph, _properties);
+
+    auto executable = _graphviz_config.layout;
+    auto format = _graphviz_config.format;
+
+    auto cmd = executable + " -T" + format + " " + graph_filename + " > " + img_filename;
```

Comment:
This seems to just be a naming issue. The file is stored in some unified way that we are calling `.dot`. This could also be called `.gv` or `.graph`. The `layout` here actually refers to the executable that renders a graph in the classic `dot` layout or in a circular `circo` layout for example. AFAIK the layout is only determined by the executable. (to be honest I stole this from @mrzzzrm, maybe he has more insight)